StdIO + runCommand (with CaptureIO)

ArielAbreu opened the pull request:
I believe I got it working WITHOUT conflicts. I resolved a couple (if not all) of the issues commented on in  #60, and took into account the possible improvements. @kesla I removed the callback from `runCommand`, it now just returns the events so you can load it into a variable and use it from there. @iefserge I made a `StdIO` class, with a `CaptureIO` module for `runCommand`, with `runtime.stdio` being an instance of `StdIO`. I also pushed a commit that creates a `stdio` global and makes `runtime.stdio` point to it.

EDIT: I just pushed a commit that adds `stdnet`, and splits stdio up. Now there's 4 new globals:
  * `stdout`
  * `stdin`
  * `stdnet`
  * and `stdio`
...instead of just stdio.

I think this is better since you can have multiple functions with the same name in different `io`s (ex: `write` in both `stdout` and `stdnet`).

ArielAbreu commented:
Wait, gotta fix it.

kesla commented:
What's the reasoning behind stdnet? Also, @iefserge do we really want runtime-core to use node.js-http? To me it feels like runtime should be lower level than that

ArielAbreu commented:
@kesla Well, the idea was that since runtime.js doesn't have any filesystem (@iefserge said he was going to remove VFS), it should be supported by default to download files, and later store them into buffers. stdnet is going to do that, although, now that you mention it, maybe it should be an optional module as well. Yeah, I'll put it into a module.

kesla commented:
@ArielAbreu If you use runtimeify you get a working `require('http')` so I then don't understand what the value is to wrap it into core. So I guess it depends on how you define "by default". I think that you should be able to download files using runtimeify, but not by only using runtimejs.

ArielAbreu commented:
@kesla Yes, but if I create a module, I can add it to runtimeify's builtins and do what you mean, using runtimeify to allow downloading files.

ArielAbreu commented:
@kesla @iefserge I created `runtime-utils`, a module containing utils for `runtimejs`. Right now, it's only util is `stdnet`.

iefserge commented:
I think runtime-utils should not be necessary, we should be able to use https://www.npmjs.com/package/request for file download.

ArielAbreu commented:
Well, yes. I'll remove the stdnet from runtime-utils. But, it listens for 'end', which doesn't work because there is no `destroy()` function in `runtime-node-net`. I created a PR in `runtime-node-net` that fixes this which I need @kesla to approve.

iefserge commented:
@kesla yes, I think node compatibility should be provided as an extra module(s).

iefserge commented on js/core/index.js:
```diff
@@ -24,6 +24,15 @@ var keyboard = require('./keyboard');
+var stdio = require('./stdio');
+
+global.stdout = new stdio.stdout();
+global.stdin = new stdio.stdin();
```
I think we shouldn't use globals, let's pass it as an argument to commands.
Curious why two objects instead of single stdio?

iefserge replied:
I think it's nice that they're separate though, we could pass them as an environment argument to commands `env = { stdin, stdout, ...other env stuff }`

iefserge replied:
let's capitalize constructor names, you can use `npm run lint` to lint js

iefserge commented on js/service/shell/captureio.js:
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+var events = require('events');
+exports.events = new events.EventEmitter();
```
I'm not sure why captureio needed. It should be easier to just make custom object with properties
```
{
 write: function() {}
 readLine: function() {}
}
```
and pass that to command to capture io

ArielAbreu replied:
CaptureIO is basically that, except in a different file so it's easier to manage. `runCommand` creates a new CaptureIO for each command because they each have properties belonging to them (ex: `readln`). PLUS, it has events that are returned to whatever called `runCommand`. That's why I made it.

iefserge commented on js/service/shell/index.js:
```diff
@@ -45,3 +46,19 @@ exports.setCommand = function(name, cb) {
   commands.set(name, cb);
 };
+
+exports.runCommand = function(name, args) {
```
How can command stdio be redirected? We could pass all input data as env for this command to use:
```js
env = {
  stdin,
  stdout,
  args
}

function(name, env, cb) {
}
```

Do we need a callback here to know when it's finished?

ArielAbreu replied:
Yes! And it should be named `env` in the callbacks in order to override the global `env` i'm adding.

iefserge commented on js/shared/input.js:
```diff
@@ -0,0 +1,123 @@
+// Copyright 2014-2015 runtime.js project authors
```
`shared` might be confusing I guess, we could move this to tty

ArielAbreu replied:
Well, it needs to be in a root level folder (root level being `js`) because it's used in `js/service/shell/input-box.js` and `js/core/stdio/stdin.js`.

iefserge replied:
`service/shell` should probably use input provided by tty stdin interface instead of input-box directly

ArielAbreu commented:
@iefserge I moved the globals into an `env` global, but `CaptureIO` is still there because it's easier to manage and fix in a separate file, plus it needs to be created with `new` so that each command has it's own special properties.

iefserge commented on js/core/stdio/stdout.js:
```diff
@@ -0,0 +1,31 @@
+    runtime.tty.print(text, 1, fg, bg);
+  }
+
+  writeln(text, fg, bg) {
```
writeln sounds not very javascripty :)
Btw, colors are probably not the part of stdio interface, expose them as a separate `setColor/setBackground` methods and make write accept any number of arguments?
Hm, now that I think about it, I think it would be nicer to make this class a pure interface. How I see it:

```js
class Stdout {
  constructor() {
    this.onwrite = function() {}
    this.onsetcolor = function(){}
    this.onsetbackgroundcolor = function() {}
  }

  write(...args) {
   this.onwrite(args.join(' '));
  }
  
  writeLine(...args) {
   this.onwrite(args.join(' ') + '\n');
  }

  setColor(fg) {
   this.fg = fg;
  }

  setBackgroundColor(bg) {
    this.bg = bg;
  }
}

class Stdin {
  // the same idea as stdout
  constructor() {
    this.onreadchar = function() {}   
  }
 
  readChar() {
    this.onreadchar(cb);
  }
 
  readLine() {
    // read the full line here using readchar
    this.onreadchar(cb);
  }
}


// Create main stdout

var defaultStdout = new runtime.tty.Stdout();  // is this part of tty?
var defaultStdin = new runtime.tty.Stdin();

defaultStdout.onsetcolor = function() {
  // change color
};
defaultStdout.onwrite = function(data) {
  runtime.tty.print(data, fg, bg);
};
defaultStdin.onreadchar = function(cb) {
  runtime.tty.readChar(cb);
}

shell.runCommand({
  stdout: defaultStdout,
  stdin: defaultStdin,
  args: {
   key: value
  }
}, function(statusCode) {
  // completed
})

// Command

setCommand(function(env, cb) {
 env.stdout.write('Output');
});

```

iefserge replied:
What do you think?

iefserge replied:
Output could be captured like:

```js
var myStdout = new runtime.tty.Stdout(); 
myStdout.onwrite = function(data) {
  console.log('captured', data);
};

runCommand({ stdout: myStdout }, cb);
```

ArielAbreu replied:
Hmm, yes. It seems more 'javascripty' than my version of the code, plus it has event handlers! Although, to capture the code, `onwrite` could `return false`, telling write not to output the line to the screen.

ArielAbreu replied:
Whoops, sorry. Hadn't noticed that all `write` does is call `onwrite`.

iefserge replied:
`onwrite` by default is not connected to a screen, only `defaultStdout` instance outputs written data to the screen
```js
defaultStdout.onwrite = function(data) { runtime.tty.print(data, fg, bg); };
```
myStdout (as in example above) could do a completely different thing. You could even pipe data from one stdout to another like
```js
myStdout.onwrite = otherStdout.onwrite;
```

iefserge replied:
`defaultStdout` could be exposed as `runtime.tty.stdout` (different from class `runtime.tty.Stdout`)

ArielAbreu replied:
I'm working on implementing all of this.

ArielAbreu commented:
@iefserge I've completely reimplemented the IO based on your comment. A little example:
```js
'use strict';

var runtime = require('runtimejs');

runtime.shell.setCommand('foobar', function(args, env, cb) {
  // get input
  env.stdin.readLine(function(text) {
    // echo the input
    env.stdout.writeLine(text);
    // Send return code 0:
    cb(0);
  });
});

var myout = runtime.stdio.createOut({
  onwrite: function(text) {
    // it's obvious what this does:
    // "hey! it's output: sometext"
    env.stdout.writeLine("hey! it's output: ", text);
  }
});

var myin = runtime.stdio.createIn({
  onreadline: function(cb) {
    // send 'blah blah text' to the callback as the input
    cb('blah blah text');
  }
});

// Run 'foobar' with 'myout' as stdout and 'myin' as stdin
runtime.shell.runCommand('foobar', {
  stdout: myout,
  stdin: myin
}, function(code, redrawPrompt) {
  // code can be 0, and...
  // 0 == false, so check for null specifically
  if (code !== null) {
    env.stdout.writeLine('exited with code ' + code);
    // redraw the prompt
    redrawPrompt();
  }
});
```

This code will output:
"hey! it's output: blah blah text"
"exited with code 0"

iefserge commented:
Looks very nice! :+1: 
I'm going to do a more detailed review soon.

ArielAbreu commented:
@iefserge Maybe the input box listener could be configured to use `runCommand()` so that it can get the exit code of the app. For anything non-zero (undefined would count as zero), a red `X` could be displayed at the beginning of the prompt (like some zsh themes do).

ArielAbreu commented on js/core/stdio/stdout.js:
```diff
@@ -0,0 +1,51 @@
+      text[i] = arguments[i];
+    }
+    this.onwrite(text.join(' '));
+  }
```
`write()` and `writeLine()` use the `arguments` object because `...args` still isn't supported.

iefserge replied:
yep

ArielAbreu commented:
I've added a feature (?) that displays a red `X` at the beginning of the prompt when the exit code is non-zero. I also fixed a bug that allows you to delete the prompt after an app uses `read` or `readline` by creating new instances of `shared/input.js` every time it's used. Also fixed another bug where the text caret would stay when using `readline` (it's still present in `read`, though).

ArielAbreu commented:
@iefserge I added a commit that catches errors in `runCommand` and passes them to the callback Node.js style. So the `runCommand` part of the example above is now:
```js
// ...
runtime.shell.runCommand('foobar', {
  stdout: myout,
  stdin: myin
}, function(err, code, redrawPrompt) {
  if (err) {
    // handle the error
  }
  // code can be 0, and...
  // 0 == false, so check for null specifically
  if (code !== null) {
    env.stdout.writeLine('exited with code ' + code);
    // redraw the prompt
    redrawPrompt();
  }
});
// ...
```

iefserge commented on js/core/index.js:
```diff
@@ -43,3 +57,44 @@ function Runtime() {
+    if (funcs) {
+      funcs.onwrite = funcs.onwrite || runtime.stdio.stdout.onwrite;
+      funcs.onsetcolor = funcs.onsetcolor || runtime.stdio.stdout.onsetcolor;
+      funcs.onsetbackgroundcolor = funcs.onsetbackgroundcolor || runtime.stdio.stdout.onsetbackgroundcolor;
```
I think we should not default those functions to stdout, it could be very unexpected. Let's just assign null by default that does nothing.

iefserge commented on js/core/index.js:
```diff
@@ -24,9 +24,23 @@ var keyboard = require('./keyboard');
+    StdIn: stdio.StdIn
+  };
+  // Copy everything from `tty` to `this.tty`
+  copy(tty, this.tty);
```
why do we need a copy?

ArielAbreu replied:
Sorry, i'd added more stuff and then I decided to delete them, guess I forgot about the copy.

iefserge commented on js/core/index.js:
```diff
@@ -43,3 +57,44 @@ function Runtime() {
+  stdin: runtime.tty.stdin,
+  StdIn: runtime.tty.StdIn,
+  StdOut: runtime.tty.StdOut,
+  createOut: function(funcs) {
```
I'm not sure if we should support convenience functions, let's just provide a raw building blocks for other libraries or streaming interfaces to build on top. It might be confusing to figure out the difference between `new StdOut` and `createOut`. What do you think?

ArielAbreu replied:
I guess it could be removed.

iefserge commented on js/core/index.js:
```diff
@@ -43,3 +57,44 @@ function Runtime() {
+  }
+};
+
+global.env = {
```
probably not good to expose it as global, it's accessible as `runtime.stdio` anyway

iefserge commented on js/service/shell/index.js:
```diff
@@ -20,6 +20,41 @@ var runtime = require('../../core');
+  opts.stdout = opts.stdout || env.stdout;
+  opts.stdin = opts.stdin || env.stdin;
+
+  try {
```
try catch wouldn't be able to catch async errors, so might be confusing that sometimes it works (error in sync code) and sometimes not (error in async code). I'd remove it to make it behave the same way

iefserge commented on js/service/shell/index.js:
```diff
@@ -20,6 +20,41 @@ var runtime = require('../../core');
+
+  var stringargs = opts.args.join(' ');
+
+  opts.stdout = opts.stdout || env.stdout;
```
let's use `runtime.stdio.stdout` instead of global

iefserge commented on js/core/index.js:
```diff
@@ -43,3 +57,44 @@ function Runtime() {
+
+runtime.tty.stdout = new runtime.tty.StdOut();
+runtime.tty.stdin = new runtime.tty.StdIn();
+initio(runtime);
```
let's move 
```
runtime.tty.stdout = new runtime.tty.StdOut();
runtime.tty.stdin = new runtime.tty.StdIn();
```
into initio

iefserge commented:
@ArielAbreu hey, I left some somments, overall it looks very nice

iefserge commented on js/core/index.js:
```diff
@@ -24,9 +24,23 @@ var keyboard = require('./keyboard');
-  this.tty = tty;
+  var self = this;
+  this.tty = {
+    StdOut: stdio.StdOut,
```
easier to type i guess :)

ArielAbreu commented:
@iefserge I think I fixed everything (correct me if i'm wrong). I also removed the `try` and `catch`, but there needs to be a way to stop the function and prevent errors from being thrown, otherwise, the VM has to be restarted.

iefserge commented on js/core/stdio/stdout.js:
```diff
@@ -0,0 +1,51 @@
+    this.onsetcolor = function() {};
+    this.onsetbackgroundcolor = function() {};
+    this.fgcolor = runtime.tty.color.WHITE;
+    this.bgcolor = runtime.tty.color.BLACK;
```
this variables are not being used

ArielAbreu replied:
Which, the colors? They are used in default io.

iefserge replied:
i see they're used in initio, but Stdout should probably not care about colors itself but just call onsetcolor/onsetbackgroundcolor to the system that actually manages colors (usually tty)

iefserge commented on js/core/stdio/index.js:
```diff
@@ -0,0 +1,21 @@
+var stdin = require('./stdin');
+
+module.exports.StdOut = stdout;
+module.exports.StdIn = stdin;
```
this probably should be named Stdin/Stdout as well

iefserge commented on js/core/stdio/stdin.js:
```diff
@@ -0,0 +1,48 @@
+
+'use strict';
+
+class StdIn {
```
let's name consistently as Stdin

iefserge commented on js/index.js:
```diff
@@ -36,14 +36,14 @@ runtime.dns = require('./service/dns-resolver');
-runtime.shell.setCommand('1', function(args, cb) {
-  runtime.tty.print('OK.\n');
+runtime.shell.setCommand('1', function(args, env, cb) {
+  env.stdout.writeln('OK.');
```
writeln?

ArielAbreu replied:
Forgot to update that.

iefserge commented on js/core/index.js:
```diff
@@ -24,9 +24,13 @@ var keyboard = require('./keyboard');
 function Runtime() {
   this.tty = tty;
+  this.tty.Stdout = stdio.Stdout;
+  this.tty.Stdin = stdio.Stdin;
```
this could be removed because we already expose Stdin/Stdout in runtime.stdio

ArielAbreu replied:
Ok

iefserge commented on js/core/initio.js:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  self.tty.stdout.onsetcolor = function(fg) {
+    self.tty.stdout.fgcolor = fg;
```
fgcolor / bgcolor could be managed here by initio.js instead of being a part of stdout

ArielAbreu commented:
Guess I forgot quite a few things 😁

ArielAbreu commented:
@iefserge I think I fixed the problems, plus I also fixed the shell commands.

iefserge commented:
@ArielAbreu tried to merge, I get 
```
Uncaught exception: /bundle.js:473: TypeError: Cannot read property 'Stdout' of undefined
TypeError: Cannot read property 'Stdout' of undefined
    at Object.7 (/bundle.js:473:37)
    at s (/bundle.js:1:254)
    at /bundle.js:1:305
    at Object.<anonymous> (/bundle.js:425:14)
    at Object.6../initio (/bundle.js:455:4)
    at s (/bundle.js:1:254)
    at /bundle.js:1:305
    at Object.<anonymous> (/bundle.js:5373:15)
    at Object.59../core (/bundle.js:5405:4)
    at s (/bundle.js:1:254)
```
`npm run lint` outputs some minor code-style issues.

iefserge commented:
I can take care of this if you'd like.

ArielAbreu commented:
Sorry, just got back. I'm gonna take a look.

ArielAbreu commented:
@iefserge I fixed the undefined issue, had to move the code in `initio` outside the exported function <i>into</i> the exported function. I was being run before `runtime` was defined, that's why it was giving an error. I also fixed the `new runtime.tty.Stdout()` part, since `Stdout()` is now in `runtime.stdio`.

ArielAbreu commented:
@iefserge I also tried to fix some issues with `lint`, but I got a merge conflict, so I decided not to do that now (since it's only minor, plus the code works without it).

piranna commented:
Just a question/sugestion: should be the stdio, stdout & stderr concepts moved out of the tty, and maybe also from runtime.js core concepts? In fact, the tty only understand about input stream (keyboard) and output stream (screen), so the diferences between stdout and stderr are only software abstractions and could be at the same level of the colors management...

My two cents.

ArielAbreu commented:
Maybe. Note: right now, there is no `stderr` (maybe I should add it, not sure).

piranna commented:
Ok :+1:

iefserge commented:
@piranna I think there needs to be some interface in the core that could be used by plugins (pluggable shells, REPLs, they could possibly interact with each other). It could also be theoretically used for other kinds of input/output in the core.
 `stderr` could be easily provided by another Stdout object, not sure if it's necessary though.

piranna commented:
> stderr could be easily provided by another Stdout object, not sure if it's necessary though.

Python implement stdio as a File object that when read return the data from the keyboard and when written put in in the screen (similar to a Node.js Duplex stream object), and all stdin, stdout and stderr are references to the same File object, so stdout and stderr are the same.

I think the core should be as minimalist as possible, mostly a wrapper over the hardware. From an API perspective, there could be only one write stream and stdout could accept only string and buffer objects while stderr could capture the Error objects written to it, this way you only have a stream with several kinds of data instead of several streams. I think it's easier to manage...

iefserge commented:
@piranna @ArielAbreu hm, I like the idea to merge stdout and stdin into one Stdio object to simplify things. `write/writeLine` functions convert all input to strings. We could add `writeError` specifically for errors and let the receiver figure out what to do with them (tty could print error.stack using red color, for example).

ArielAbreu commented:
What do you mean "convert all input to strings". Isn't write output all they do? What do they have to do with input? (but, again, maybe, first I need a little idea of what you mean).

iefserge commented:
@ArielAbreu this is how it works right now, I think it's nice that `write` converts any arguments to strings.

ArielAbreu commented:
I thought you meant user input, my bad.

iefserge commented:
@ArielAbreu I'd say let's merge Stdout and Stdin into one Stdio object (rename and expose class as `runtime.stdio.StdioInterface` and default `stdio` object that interacts with tty as `runtime.stdio.defaultStdio`). Other than that, this looks good!

ArielAbreu commented:
I'll combine them.

ArielAbreu commented:
@iefserge I've combined `Stdout` and `Stdin` into one `Stdio` class available as `runtime.stdio.Interface`, with a default io as `runtime.stdio.defaultio` (with `runtime.tty.io` as an alias). It also has a new `writeError()` method that accepts either a string or an `Error` object and prints it's `stack` property.

I've also been thinking there should be a way of sending data between applications. Maybe a global event object could be exposed that allows events to sent to applications like:
```js
runtime.events.emit('appname-someevent', mydatavar);

// someapp
runtime.shell.setCommand('appname', function(args, env, cb) {
  env.on('someevent', function(datavar) {
    // i don't know, do something with `datavar`
  });
});
```
I don't know, maybe i'm getting ahead of myself. 😐 Just a random idea.

iefserge commented:
@ArielAbreu Very nice! However I have a few comments related to API design.

1.The issue with calling it `Interface` is that you cannot simply do
```js
obj instanceof Interface
```
to check the object type and I think it's very confusing that class name is different from the property name visible to user. For example, there is `runtime.net.TCPSocket`, class named `TCPSocket` and `obj instanceof TCPSocket` works correctly. I think we should follow the same rule across the API.

2.Let's remove the alias, I would argue that this creates unnecessary confusion. I'd say we should have every object exposed no more than once.

3.`defaultio` consists of two words, it should probably be named either `defaultIO` or maybe just `defaultStdio`, because `defaultIO` could refer to different kinds of io, not just "standard IO". Simply `default` would also work (is this a good idea to use a js keyword though?). I think we should either call it `io` or `stdio` everywhere consistently (`env` property as well). Please, let's call it stdio because io is very vague :)

I'm sorry, I should've documented the rules currently used by other APIs somewhere so we can make sure it's consistent. We can change these rules of course, but we would have to update all other APIs as well.

`writeError` is nice! :+1: 
Let's think about events and communication separately from this feature, this could get very complicated.

ArielAbreu commented:
@iefserge I've fixed the issues. Sorry about the API, I guess everyone has their own way of doing things.

piranna commented:
> I guess everyone has their own way of doing things

Absolutely :-D Me, I would have splitted the screen and the keyboard from the stdio and would have implemented it as an object combining them :-)

iefserge commented:
@ArielAbreu Great work!! Merged!

iefserge commented:
@piranna that's pretty much what `StdioInterface` is for, it's in no way connected to a screen and keyboard. `defaultStdio` is the instance that implements this interface using screen/kbd :)

piranna commented:
> @piranna that's pretty much what StdioInterface is for, it's in no way connected to a screen and keyboard. defaultStdio is the instance that implements this interface using screen/kbd :)

Ah cool! :-D
